### PR TITLE
The solution of Time Converion Challenge

### LINKED ISSUE #36: Challenge_easy_time_conversion
Write a function that converts a given time from EST to AST. The function should take the input time in EST in the format HH:MM (24-hour format) and return the equivalent time in AST in the same format.

Example: 
Input: 
12:30
Output:
20:30

### PR DESCRIPTION
**Pull request by safaabuzaid**
---
name: solution review 
about: Time Conversion Function 
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/Test_Time_Conversion.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+"""
+Test module for alternate_elements function.
```
Comment (Kareiman) [inline solutions/tests/Test_Time_Conversion.py]: This line talks about a different function
yours is time_conversion not alternate_elements

File: solutions/tests/Test_Time_Conversion.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+Contains intentionally buggy tests for debugging practice.
+
+Test categories:
+    - Standard cases: typical lists with different lengths
+    - Edge cases: empty lists, single elements
+    - Defensive tests: wrong input types, assertions
+
+Created on 2024-12-06
+Author: Claude AI
```
Comment (Kareiman) [inline solutions/tests/Test_Time_Conversion.py]: The author's name is incorrect

File: solutions/Time_Conversion.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    Takes input time in the format HH:MM (24-hours format) and returns
+    the equavelent time in AST in the same format.
+
+    Parameters:
+        EST_Time (str): time in EST in the format 24 hour
+
+    Returns -> str: time in AST in the format 24 hour
+
+    Raises:
+        AssertionError: if the input is not a string
```
Comment (Kareiman) [inline solutions/Time_Conversion.py]: The errors could be:
ValueError in case the entry was outside the expected range for each of the hours, minutes
TypeError: if entry wasn't a string

File: solutions/tests/Test_Time_Conversion.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        """It should raise AssertionError for hours are out of range"""
+        with self.assertRaises(AssertionError):
+            Time_Conversion("25:30")
+
+    def test_minutes_out_of_range(self):
+        """It should raise AssertionError for minutes are out of range"""
+        with self.assertRaises(AssertionError):
+            Time_Conversion("12:69")
+
+    def test_strings(self):
```
Comment (Kareiman) [inline solutions/tests/Test_Time_Conversion.py]: add more tests (boundary tests) where entry was text and letters or mix between numbers and letters

File: solutions/tests/Test_Time_Conversion.py
Comment (Kareiman) [inline solutions/tests/Test_Time_Conversion.py:1]: I believe the name should be test_Time_Conversion.py with a small t for test

File: solutions/Time_Conversion.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+
+def Time_Conversion(EST_Time: str) -> str:
+    """Returns a time in Arabia Standard Time zone (AST)
+
+    Takes input time in the format HH:MM (24-hours format) and returns
+    the equavelent time in AST in the same format.
+
+    Parameters:
+        EST_Time (str): time in EST in the format 24 hour
```
Comment (Kareiman) [inline solutions/Time_Conversion.py:17]: variables should be named in snake_case, so instead of EST_Time, you can write est_time

**Comment by Kareiman** [comment]
Hello Safaa,
You did a great job, only minor fixes and you will pass the whole checklist!
I think the tricky part is about the Assertions and Error types, take a look at my suggestions and let me know what do think.

**Review by Kareiman — CHANGES_REQUESTED** [review]
Let me know if you can see my comments now

File: solutions/tests/Test_Time_Conversion.py (hunk 4 above)
Comment (safaabuzaid) [reply solutions/tests/Test_Time_Conversion.py]: Ok, I will

**Comment by safaabuzaid** [comment]
Thank you @Kareiman for your review, I updated my code according to your guides, so please check again and share with me your feedback.

**Comment by Kareiman** [comment]
> Thank you @Kareiman for your review, I updated my code according to your guides, so please check again and share with me your feedback.

Impressive work on the code @safaabuzaid 
Can you also push the changes you made on the unit-test file.

File: solutions/tests/test_Time_Conversion.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+            Time_Conversion("25:30")
+
+    def test_minutes_out_of_range(self):
+        """It should raise ValueError for minutes are out of range"""
+        with self.assertRaises(ValueError):
+            Time_Conversion("12:69")
+
+    def test_strings(self):
+        """It should return the time conversion properly"""
+        self.assertEqual(Time_Conversion("12:12"), "20:12")
```
Comment (Kareiman) [inline solutions/tests/test_Time_Conversion.py]: This is the typical scenario, adding another one would be helpful.
 You have also defensive cases, but you need to add edge cases, like testing boundaries or limits of the input, ensuring the function behaves correctly at or near these extremes.
 Organizing them accordingly would be helpful

File: solutions/tests/Test_Time_Conversion.py (hunk 4 above)
Comment (Kareiman) [reply solutions/tests/Test_Time_Conversion.py]: apologies for not making this clear, this example I provided "mix entry" presents a defensive case
boundary cases should be: boundary values within the input range or unusual but valid inputs.

Comment (safaabuzaid) [reply solutions/tests/Test_Time_Conversion.py]: Thank you for clarification, please have a look to the last update.

File: solutions/tests/test_Time_Conversion.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+    # Edge cases
+
+    def test_equal_minutes_hours(self):
+        """It should return the time conversion properly"""
+        self.assertEqual(Time_Conversion("12:12"), "20:12")
+
+    def test_minimum_time(self):
+        """It should return the time conversion properly"""
+        self.assertEqual(Time_Conversion("00:00"), "8:0")
```
Comment (Kareiman) [inline solutions/tests/test_Time_Conversion.py]: According to the format the result should be "08:00"

File: solutions/tests/test_Time_Conversion.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+    def test_maximum_time(self):
+        """It should return the time conversion properly"""
+        self.assertEqual(Time_Conversion("23:59"), "7:59")
```
Comment (Kareiman) [inline solutions/tests/test_Time_Conversion.py]: According to the format the result should be "07:59"

File: solutions/tests/test_Time_Conversion.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+Author: Safaa Osman
+"""
+
+import unittest
+from ..Time_Conversion import Time_Conversion
+
+
+class TestTimeConversion(unittest.TestCase):
+    """Test suite for the Time Conversion function - contains buggy tests!"""
+
```
Comment (Kareiman) [inline solutions/tests/test_Time_Conversion.py]: Regular cases can be written instead of boundary

**Review by Kareiman — APPROVED** [review]
Great job Safaa.
Everything looks solid, and the code is clean and easy to follow. 
I learned a lot from you in this process and I enjoyed working with you 
Keep up the excellent work!
